Explorer: Support class extension

Pixep opened the pull request:
Support class extension with `extends`
Relates to #1881 

Add support for:
* Class extension with `extends`
* Using base class methods and members

Pixep committed:
Explorer: add basic support for 'extends' and calling base class fn

Pixep committed:
Explorer: rename & chg type for 'extends' in class declaration

Pixep committed:
Explorer: add support for parent class member variables with 'extends'

Pixep committed:
Explorer: minimize code duplication to check class members

Pixep committed:
explorer: error when shadowing base class members

Pixep committed:
Merge branch 'trunk' into feature/support-class-extension

Pixep committed:
Merge branch 'trunk' into feature/support-class-extension

Pixep commented:
Opening for feedback on the code, and from an architecture standpoint, as I suspect changes might be needed to make sure we drive this in the proper direction.

Pixep commented:
Adjusted reviewers, as I suppose we need to bother Chandler with that just yet :). Feel free to update.

jonmeow commented on explorer/testdata/class/fail_extends_non_class.carbon:
```diff
@@ -8,13 +8,10 @@
-  fn F() {}
-// CHECK:COMPILATION ERROR: {{.*}}/explorer/testdata/class/fail_class_extends.carbon:[[@LINE+1]]: Class extension with `extends` is not supported yet
+// CHECK:SYNTAX ERROR: {{.*}}/explorer/testdata/class/fail_extends_non_class.carbon:[[@LINE+1]]: syntax error, unexpected sized_type_literal, expecting identifier
+class C extends i32 {
```
```suggestion
class C extends 3 {
```

If you want something that's guaranteed not to work, it'd be better to do something like `extends 3`... `i32` should eventually translate to a normal type, and so extending it may fail (i.e., because `i32` probably will be final) but it should be a class.

Pixep replied:
Done

jonmeow commented on explorer/testdata/class/fail_class_extends_overrides_member.carbon:
```diff
@@ -0,0 +1,24 @@
+}
+
+class D extends C {
+  // CHECK: COMPILATION ERROR: {{.*}}/explorer/testdata/class/fail_class_extends_overrides_member.carbon:[[@LINE+1]]: A member named `var_c` already exists in this class or its parents
```
Generally it's fine if explorer errors are a little rough, but can this point to the conflicting definition?

Pixep replied:
Dropped entirely for now, following conversation re: shadowing

jonmeow commented on explorer/syntax/parser.ypp:
```diff
@@ -1132,8 +1132,8 @@ class_declaration_extensibility:
     { $$ = std::nullopt; }
-| EXTENDS expression
-    { $$ = $2; }
+| EXTENDS identifier
```
I think `expression` was correct, here. Consider generic types such as `MyContainer(ElementType)`: these are expressions, not identifiers. This change probably shouldn't modify parser.ypp; it can change the checking of the extended type to be restrictive, and support for extending generic types can be a TODO, but it's undesirable to go to an implementation that constrains to an identifier.

Regardless of whether generic types are a TODO, they should also be tested; either to be working or a "not supported yet" placeholder error, depending.

Note, I believe this affects changes to several files here, so the rest of my comments will be light.

Pixep replied:
Done

jonmeow commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -3443,6 +3504,23 @@ auto TypeChecker::TypeCheckClassDeclaration(
     CARBON_RETURN_IF_ERROR(TypeCheckDeclaration(m, class_scope, class_decl));
     CARBON_RETURN_IF_ERROR(CollectMember(class_decl, m));
   }
+  std::unordered_map<std::string_view, SourceLocation> member_names;
+  const auto check_duplicate_members =
+      [&member_names](Nonnull<Declaration*> m) -> ErrorOr<Success> {
+    if (const auto name = GetName(*m); name.has_value()) {
+      if (auto [_, inserted] = member_names.insert(
+              std::make_pair(name.value(), m->source_loc()));
+          !inserted) {
+        return ProgramError(m->source_loc())
+               << "A member named `" << name.value()
+               << "` already exists in this class or its parents";
+      };
+    }
+    return Success();
+  };
+  CARBON_RETURN_IF_ERROR(
+      VisitAllClassMembers(*class_decl, check_duplicate_members));
```
Right now our duplicate name checking is in StaticScope; I'm hesitant to add more.

So for example, having the a function in a parent class be shadowed by a child class may be valid in C++: https://cpp.compiler-explorer.com/z/Yb9zooMn7

There are reasons to want to allow this -- it allows adding functionality to the base class without worrying about breaking inheritors. Glancing at [the design](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/classes.md#name-lookup-in-member-function-definitions), I don't see this being ruled out, so I think it should probably be allowed until there's a clearer decision.

Pixep replied:
Thanks for the ref to StaticScope.

Supported in C++ yes, but for Carbon the way I read it, we don't want to allow child classes to non-explicitly (override or other) hide a method (making it expressive, not pitfalls, etc.). I'm thinking about this reference: https://github.com/carbon-language/carbon-lang/blob/8e8e8ced6014ae131290a7eb74631cfa58c80c4f/docs/design/classes.md?plain=1#L1906

That said this reference targets member functions specifically. I think I remember something more generic including member variables in the class design, though I'd have to dig a bit deeper (if it does, in fact, exist).

Maybe the intent needs to be clarified too.
Do you have a different interpretation?

jonmeow replied:
Thanks, I'd missed that. But note it's under ["Future work"](https://github.com/carbon-language/carbon-lang/blob/8e8e8ced6014ae131290a7eb74631cfa58c80c4f/docs/design/classes.md?plain=1#L1784) -- that means it's been thought about, but not in enough depth to make a decision.

jonmeow replied:
Also note the paragraph before the bullets:

> Beyond that, we may want some rules or restrictions about defining methods in a
> derived class with the same name as a base class method without overriding it.
> There are some opportunities to improve on and simplify the C++ story:

I would emphasize "may". While the wording in the bullets is strong, it's wrapped by a tentative statement.

Anyways, I wanted to add that even if we _do_ add this, the approach should change: in addition to StaticScope, mixins need to be considered, and it feels like the current approach is less suitable for access control (also maybe name lookup and overloads?). Since this is undecided, my thought's really that it may be better to build out the other pieces (particularly access control) and that'd probably help shape how name conflicts would be managed if the leads decide to head in that direction.

Pixep replied:
Done, dropped as not aligned with the spec.

jonmeow reviewed (commented):
Regarding chandlerc, he can review changes, no need to remove him as reviewer. I can still review here, though.

That said, I think there are a couple key issues here: one about not preventing extending generics, and another about not preventing name conflicts between parents and children. I think these will probably cause significant changes so I'm minimizing style comments on this PR.

I think the things like LookupFunction are probably good and the right direction, but I'm not sure how they'll be affected by the rest.

Pixep commented:
> That said, I think there are a couple key issues here: one about not preventing extending generics, and another about not preventing name conflicts between parents and children. I think these will probably cause significant changes so I'm minimizing style comments on this PR.

Thanks! I appreciate the directions, that's exactly what I was looking for.

Pixep committed:
Explorer: restore base class to 'Expression' for parser

Pixep committed:
Explorer: drop unique members type check

Remove dependency function VisitAllClassMembers and GetAllClassMembers

Pixep committed:
Explorer: add more class inheritance tests

Pixep committed:
Explorer: check that base class has `base` class prefix

Pixep commented:
- Added back chandlerc as suggested.
- Added a few more tests and `base` keyword support
- Known limitation: Base destructors are not called. I think it would probably make sense to include that in this PR as a class extension MVP, but I yet have to identify how to enable that.

Pixep commented on explorer/testdata/class/class_inheritance_shadow.carbon:
```diff
@@ -0,0 +1,42 @@
+
+fn Main() -> i32 {
+  // Initialize derived value first, base value second
+  var d: D = {.value = 2, .value = 1};
```
The current way to assign base class values when shadowed seems wrong to me. The [class design doc](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/classes.md#constructors) discusses the use of `{.base = ...` for base class initialization in constructors, but this isn't supported yet AFAIS.

I couldn't find a mention of this particular situation in the design.

This example still illustrates variable shadowing, and the fact that `me` represents the object. and not the base class, but we might likely want to not allow that in some future work

jonmeow replied:
`base =` would probably need to have support added as part of `extends` but is probably the right direction. I suspect this example actually shouldn't work -- the shadowing should be an ambiguous name lookup (I think this is the direction that function calls will go too).

Note, something that might be okay/supportable would be: `{.(C.value) = 2, .(D.value) = 1};`

But that kind of name specification is also how ambiguous functions can be resolved.

Pixep replied:
Agree about the initialization.

Regarding the `base =`, are you saying that something that you'd want to include in this PR, or changing that test to not use this form?

jonmeow commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -3410,19 +3379,37 @@ auto TypeChecker::DeclareClassDeclaration(Nonnull<ClassDeclaration*> class_decl,
+               << class_decl->name()
+               << "`. Only simple classes are currently supported as base "
+                  "class.";
+        break;
```
```suggestion
```

Pixep replied:
Yep... my bad. Done.

jonmeow commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -3357,13 +3379,37 @@ auto TypeChecker::DeclareClassDeclaration(Nonnull<ClassDeclaration*> class_decl,
+                 << "Base class `" << base_class.value()->declaration().name()
+                 << "` is `final` and cannot inherited. Add the `base` or "
+                    "`abstract` class prefix to `"
+                 << base_class.value()->declaration().name()
```
```suggestion
                 << base_type
```

I think the Value will yield a cleaner result here

Pixep replied:
This (unfortunately) prints `typeof(class C)` instead of `C`. Would have been cleaner!

jonmeow commented on explorer/interpreter/type_checker.cpp:
```diff
@@ -3357,13 +3379,37 @@ auto TypeChecker::DeclareClassDeclaration(Nonnull<ClassDeclaration*> class_decl,
+        if (base_class.value()->declaration().extensibility() ==
+            ClassExtensibility::None) {
+          return ProgramError(class_decl->source_loc())
+                 << "Base class `" << base_class.value()->declaration().name()
```
```suggestion
                 << "Base class `" << base_type
```

Pixep replied:
Same

> This (unfortunately) prints typeof(class C) instead of C. Would have been cleaner!

jonmeow reviewed (approved):
Thanks! Just a few small things to clean up.

I think the destructor support is fine not to address. There's also some work being done on #2255 that it may be better not to overlap with too much.

Pixep committed:
Explorer: Remove superfluous 'break' statement

Pixep commented:
> Thanks! Just a few small things to clean up.
> 
> I think the destructor support is fine not to address. There's also some work being done on #2255 that it may be better not to overlap with too much.

Great! A couple things I just ran into:
- I'm not sure about the naming I used for `LookupFunction` and `LookupMember`, I suppose `FindFunctionInClasses` or similar would be better for symmetry with `FindFunction` and `FindMember`. `...WithinClassHierchy` would be the clearest, but a bit verbose :). We can stick with what we have too.
- Merging trunk, I'm having a little issue with `TypeChecker::FindMixedMemberAndType` and the fact it now takes `SourceLocation source_loc`, though never makes use of it, and only returns a default `{std::nullopt}` error. The compiler isn't complaining because it recurses passing the same (unused parameter). I'd suggest dropping it, given it looks unused and outside of this function responsibility. The alternative means carrying it around in `LookupMember`, and it doesn't look great. I might have to convert it to return an `ErrorOr<>` too to preserve the current behavior of CARBON_ASSIGN_OR_RETURN around `type_checker.cpp:2206` or drop it entirely and pass a flatten list of all members.

jonmeow commented:
> * I'm not sure about the naming I used for `LookupFunction` and `LookupMember`, I suppose `FindFunctionInClasses` or similar would be better for symmetry with `FindFunction` and `FindMember`. `...WithinClassHierchy` would be the clearest, but a bit verbose :). We can stick with what we have too.

Either way, each option seems fine. Maybe "FindFunctionWithParents" if you're looking for maybe incrementally less verbose?

> * Merging trunk, I'm having a little issue with `TypeChecker::FindMixedMemberAndType` and the fact it now takes `SourceLocation source_loc`, though never makes use of it, and only returns a default `{std::nullopt}` error. The compiler isn't complaining because it recurses passing the same (unused parameter). I'd suggest dropping it, given it looks unused and outside of this function responsibility. The alternative means carrying it around in `LookupMember`, and it doesn't look great. I might have to convert it to return an `ErrorOr<>` too to preserve the current behavior of CARBON_ASSIGN_OR_RETURN around `type_checker.cpp:2206` or drop it entirely and pass a flatten list of all members.

I see:

```
          CARBON_ASSIGN_OR_RETURN(
              auto type_member, FindMixedMemberAndType(
                                    access.source_loc(), access.member_name(),
                                    t_class.declaration().members(), &t_class));
          if (type_member.has_value()) {
...
          } else {
            return ProgramError(e->source_loc())
                   << "class " << t_class.declaration().name()
                   << " does not have a field named " << access.member_name();
          }
```

@DarshalShetty is experimenting with mixins. I agree source_loc isn't used by the current code, and it may be okay to drop as a parameter. Alternately, FindMixedMemberAndType is returning ErrorOr, even though it doesn't presently return an error; the ProgramError could be returned instead of nullopt.

Getting rid of the source_loc parameter may be the better choice for you just because you aren't returning an Error right now, although note if LookupMember needs to distinguish due to failure of not finding something versus failure of finding _ambiguous_ things, it seems likely to need to return an error in the future.

Pixep committed:
Merge branch 'trunk' into feature/support-class-extension

Pixep committed:
Explorer: rename Lookup* functions to Find*WithParents

Pixep commented:
> Getting rid of the source_loc parameter may be the better choice for you just because you aren't returning an Error right now, although note if LookupMember needs to distinguish due to failure of not finding something versus failure of finding _ambiguous_ things, it seems likely to need to return an error in the future.

Sounds good, changes applied. I think this is acceptable too for now.
We have the new `TypeOf*Type` changes that I have to merge, then we should be in good shape.

jonmeow reviewed (approved):
Still LGTM, sorry about all the merges you're running into.

Pixep committed:
Merge branch 'trunk' into feature/support-class-extension

chandlerc commented:
(@jonmeow and others are definitely better reviewers for the explorer, just cleaning out my queue.)

Pixep committed:
Explorer: adapted design around removal of TypeOfClassType

jonmeow commented:
Looks great! I think this is a good state to get merged.
